GrandCentralBoard/Widgets/Bonus/Source/PeopleWithBonususFetchController.swift

```diff
@@ -0,0 +1,134 @@
+//
+
+import Foundation
+import GrandCentralBoardCore
+
+
+private let kPreferredPeopleCount = 10
+
+enum PeopleWithBonususFetchControllerError : ErrorType {
+    case IncorrectEmailAddress
+    case Cancelled
+    case Unknown
+}
+
+
+final class PeopleWithBonususFetchController {
+
+    private let requestSender: RequestSender
+
+    init(requestSender: RequestSender) {
+        self.requestSender = requestSender
+    }
+
+    func fetchPeopleWithBonuses(completionBlock: (Result<[Person]>) -> Void) {
+        fetchPeopleWithBonuses(startingFromDate: NSDate.init(), fetchedBonuses: [], completionBlock: { [weak self] result in
+            guard let strongSelf = self else { return }
+
+            switch result {
+            case .Success(let people):
+                strongSelf.fetchAvatarsForPeople(people, completionBlock: { result in
+                    switch result {
+                    case .Success(let people):
+                        completionBlock(.Success(people))
+                    case .Failure(let error):
+                        completionBlock(.Failure(error))
+                    }
+                })
+            case .Failure(let error):
+                completionBlock(.Failure(error))
+            }
+
+        })
+    }
+
+    private func fetchPeopleWithBonuses(startingFromDate date: NSDate = NSDate.init(), fetchedBonuses: [Bonus], completionBlock: (Result<[Person]>) -> Void) {
+        let take = 100
+        let requestTemplate = TimestampableRequestTemplate<BonusesRequestTemplate>.init(requestTemplate: BonusesRequestTemplate(), date: date, take: take)
+        
+        requestSender.sendRequestForRequestTemplate(requestTemplate) { [weak self] result in
+            guard let strongSelf = self else {
+                completionBlock(.Failure(PeopleWithBonususFetchControllerError.Cancelled))
+                return
+            }
+
+            switch result {
+            case .Success(let bonuses):
+                var allBonuses: [Bonus] = fetchedBonuses.reverse()
+                allBonuses.appendContentsOf(bonuses)
+
+                let people = allBonuses.reduce(Set<Person>(), combine: { people, bonus in
+                    if people.count >= kPreferredPeopleCount {
+                        return people
+                    } else {
+                        var mutablePeople = people
+                        var receiver = bonus.receiver
+                        receiver = receiver.copyWithLastBonusDate(bonus.date)
+                        mutablePeople.insert(receiver)
+                        return mutablePeople
+                    }
+                })
+
+                if people.count >= kPreferredPeopleCount || bonuses.count < take {
+                    completionBlock(.Success(Array(people)))
+                } else if let lastBonus = bonuses.last {
+                    strongSelf.fetchPeopleWithBonuses(startingFromDate: lastBonus.date, fetchedBonuses: allBonuses, completionBlock: completionBlock)
```

Comment:
I think it would be better not to call `fetchPeopleWithBonuses ` again if we do not have `people.count >= kPreferredPeopleCount` and just live with that there are less bubbles than desired.